apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/TransactionManager.java

```diff
@@ -112,38 +110,83 @@ private EntityManager getOrCreateEntityManager() {
   }
 
   /**
-   * Commits or rolls back the active transaction. If a commit fails, or if {@code shouldCommit} is
-   * false the current {@link EntityManager} is also closed.
+   * Commits or rolls back the active transaction. If a commit fails, or if {@code
+   * transactionException} is non-null the current {@link EntityManager} is also closed. Ensures
+   * that any exceptions thrown during completion are either thrown (if the transaction had been
+   * otherwise successful) or added to the failing exception as being suppressed.
    *
    * @param entityManager the {@link EntityManager} owning the transaction
-   * @param shouldCommit true if commit is needed, false if rollback is needed
+   * @param transactionException null if transaction was successful, otherwise the exception that
+   *     caused transaction to fail
    */
-  private void completeTransaction(EntityManager entityManager, boolean shouldCommit) {
-    boolean shouldClose = true;
+  private void completeTransaction(EntityManager entityManager, Exception transactionException) {
+    if (transactionException == null) {
+      completeSuccessfulTransaction(entityManager);
+    } else {
+      completeFailedTransaction(entityManager, transactionException);
+    }
+  }
+
+  /**
+   * Commits the transaction and clears the entity manager. If the commit fails the entity manager
+   * is closed. Ensures that any exceptions thrown during completion are properly thrown.
+   *
+   * @param entityManager the {@link EntityManager} owning the transaction
+   */
+  private void completeSuccessfulTransaction(EntityManager entityManager) {
+    RuntimeException completionException = null;
     try {
-      if (shouldCommit) {
-        entityManager.getTransaction().commit();
-        entityManager.clear();
-        shouldClose = false;
-      } else {
-        entityManager.getTransaction().rollback();
-      }
+      entityManager.getTransaction().commit();
+      entityManager.clear();
+    } catch (RuntimeException commitException) {
+      completionException = commitException;
+      throw commitException;
     } finally {
-      if (shouldClose) {
-        closeEntityManager();
+      if (completionException != null) {
+        closeEntityManager(completionException);
       }
     }
   }
 
+  /**
+   * Rolls back the transaction and closes the entity manager. Ensures that any exception during the
+   * rollback or close are added to the transaction exception as suppressed exceptions.
+   *
+   * @param entityManager the {@link EntityManager} owning the transaction
+   * @param transactionException the exception that caused transaction to fail
+   */
+  private void completeFailedTransaction(
+      EntityManager entityManager, Exception transactionException) {
+    try {
+      entityManager.getTransaction().rollback();
+    } catch (RuntimeException rollbackException) {
+      transactionException.addSuppressed(rollbackException);
```

Comment:
we'll only ever get a RuntimeExecption here ??

Reply:
Yes.  These methods do not throw any checked exceptions.